beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/forward/multipeer/chains/ThrottlingSyncSource.java

```diff
@@ -61,33 +69,64 @@ public SafeFuture<Void> requestBlocksByRange(
       final UInt64 startSlot,
       final UInt64 count,
       final RpcResponseListener<SignedBeaconBlock> listener) {
-    if (blocksRateTracker.approveObjectsRequest(count.longValue()).isPresent()) {
-      LOG.debug("Sending request for {} blocks", count);
-      return delegate.requestBlocksByRange(startSlot, count, listener);
-    } else {
-      LOG.debug(
-          "Rate limiting request for {} blocks. Retry in {} seconds",
-          count,
-          PEER_REQUEST_DELAY.toSeconds());
-      return asyncRunner.runAfterDelay(
-          () -> requestBlocksByRange(startSlot, count, listener), PEER_REQUEST_DELAY);
-    }
+    return blocksRateTracker
+        .approveObjectsRequest(count.longValue())
+        .map(
+            requestApproval -> {
+              LOG.debug("Sending request for {} blocks", count);
+              final RpcResponseListenerWithCount<SignedBeaconBlock> listenerWithCount =
+                  new RpcResponseListenerWithCount<>(listener);
+              return delegate
+                  .requestBlocksByRange(startSlot, count, listenerWithCount)
+                  .alwaysRun(
+                      () ->
+                          // adjust for slots with empty blocks
+                          blocksRateTracker.adjustObjectsRequest(
+                              requestApproval, listenerWithCount.count.get()));
+            })
+        .orElseGet(
+            () -> {
+              LOG.debug(
+                  "Rate limiting request for {} blocks. Retry in {} seconds",
+                  count,
+                  PEER_REQUEST_DELAY.toSeconds());
+              return asyncRunner.runAfterDelay(
+                  () -> requestBlocksByRange(startSlot, count, listener), PEER_REQUEST_DELAY);
+            });
   }
 
   @Override
   public SafeFuture<Void> requestBlobSidecarsByRange(
       final UInt64 startSlot, final UInt64 count, final RpcResponseListener<BlobSidecar> listener) {
-    if (blobSidecarsRateTracker.approveObjectsRequest(count.longValue()).isPresent()) {
-      LOG.debug("Sending request for {} blob sidecars", count);
-      return delegate.requestBlobSidecarsByRange(startSlot, count, listener);
-    } else {
-      LOG.debug(
-          "Rate limiting request for {} blob sidecars. Retry in {} seconds",
-          count,
-          PEER_REQUEST_DELAY.toSeconds());
-      return asyncRunner.runAfterDelay(
-          () -> requestBlobSidecarsByRange(startSlot, count, listener), PEER_REQUEST_DELAY);
-    }
+    long blobSidecarsCount =
+        maybeMaxBlobsPerBlock
+            .map(maxBlobsPerBlock -> maxBlobsPerBlock * count.longValue())
+            .orElse(0L);
+    return blobSidecarsRateTracker
+        .approveObjectsRequest(blobSidecarsCount)
+        .map(
+            requestApproval -> {
+              LOG.debug("Sending request for ~ {} blob sidecars", blobSidecarsCount);
```

Comment:
remove the `~` probably reads better

Reply:
Changed ~ to approximately since it is not exact number